### ISSUE TITLE
exp and sigmoid may cause inf.

### ISSUE THREAD

**Issue by guzy0324** [VincentStimper/normalizing-flows#47]
Can we do a `clamp` like before `z2` is returned in `AffineCoupling`?

For example:

```python
class AffineCouplingStable(Flow):
    """
    Affine Coupling layer as introduced RealNVP paper, see arXiv: 1605.08803
    """

    def __init__(self, param_map, scale=True, scale_map="exp"):
        """Constructor

        Args:
          param_map: Maps features to shift and scale parameter (if applicable)
          scale: Flag whether scale shall be applied
          scale_map: Map to be applied to the scale parameter, can be 'exp' as in RealNVP or 'sigmoid' as in Glow, 'sigmoid_inv' uses multiplicative sigmoid scale when sampling from the model
        """
        super().__init__()
        self.add_module("param_map", param_map)
        self.scale = scale
        self.scale_map = scale_map

    def forward(self, z):
        """
        z is a list of z1 and z2; ```z = [z1, z2]```
        z1 is left constant and affine map is applied to z2 with parameters depending
        on z1

        Args:
          z
        """
        z1, z2 = z
        param = self.param_map(z1)
        if self.scale:
            shift = param[:, 0::2, ...]
            scale_ = param[:, 1::2, ...]
            if self.scale_map == "exp":
                z2 = z2 * torch.exp(scale_) + shift
                z2 = z2.clamp(min=-1e6, max=1e6)
                log_det = torch.sum(scale_, dim=list(range(1, shift.dim())))
            elif self.scale_map == "sigmoid":
                scale_inv = 1 + torch.exp(-(scale_ + 2)) # 1 / sigmoid(scale_ + 2)
                z2 = z2 * scale_inv + shift
                z2 = z2.clamp(min=-1e6, max=1e6)
                log_scale = nn.functional.logsigmoid(scale_ + 2)
                log_det = -torch.sum(log_scale, dim=list(range(1, shift.dim())))
            elif self.scale_map == "sigmoid_inv":
                scale = torch.sigmoid(scale_ + 2)
                z2 = z2 * scale + shift
                z2 = z2.clamp(min=-1e6, max=1e6)
                log_scale = nn.functional.logsigmoid(scale_ + 2)
                log_det = torch.sum(log_scale, dim=list(range(1, shift.dim())))
            else:
                raise NotImplementedError("This scale map is not implemented.")
        else:
            z2 = z2 + param
            log_det = zero_log_det_like_z(z2)
        return [z1, z2], log_det

    def inverse(self, z):
        z1, z2 = z
        param = self.param_map(z1)
        if self.scale:
            shift = param[:, 0::2, ...]
            scale_ = param[:, 1::2, ...]
            if self.scale_map == "exp":
                z2 = (z2 - shift) * torch.exp(-scale_)
                z2 = z2.clamp(min=-1e6, max=1e6)
                log_det = -torch.sum(scale_, dim=list(range(1, shift.dim())))
            elif self.scale_map == "sigmoid":
                scale = torch.sigmoid(scale_ + 2)
                z2 = (z2 - shift) * scale
                z2 = z2.clamp(min=-1e6, max=1e6)
                log_scale = nn.functional.logsigmoid(scale_ + 2)
                log_det = torch.sum(log_scale, dim=list(range(1, shift.dim())))
            elif self.scale_map == "sigmoid_inv":
                scale_inv = 1 + torch.exp(-(scale_ + 2)) # 1 / sigmoid(scale_ + 2)
                z2 = (z2 - shift) * scale_inv
                z2 = z2.clamp(min=-1e6, max=1e6)
                log_scale = nn.functional.logsigmoid(scale_ + 2)
                log_det = -torch.sum(log_scale, dim=list(range(1, shift.dim())))
            else:
                raise NotImplementedError("This scale map is not implemented.")
        else:
            z2 = z2 - param
            log_det = zero_log_det_like_z(z2)
        return [z1, z2], log_det
```

**Comment by VincentStimper** [comment]
Hi @guzy0324,

thanks for this interesting suggestion. You are welcome to implement your own flow layers for `normflows` like this and try whether it works in practice. However, I will not implement it, since this breaks the invertibility of the flow. Instead, I suggest shrinking the scale parameter, i.e. by applying a `tanh` output function in the [`MLP`](https://github.com/VincentStimper/normalizing-flows/blob/d1a6af7b15bf0245f1bc4c18e1694541872b29d8/normflows/nets/mlp.py#L5) module, which will also prevent `z` from becoming too large. Using `"sigmoid"` or `"sigmoid_inv"` as `scale_map`, depending on in which direction you use the flow during training, should also help.

Best regards,
Vincent

**Comment by guzy0324** [comment]
Oh, I didn't notice that applying `clamp` on `z2` would break the invertibility. How about applying `clamp` on `scale`? It seems to prevent `z` from becoming too large and preserve the invertibility.

**Comment by VincentStimper** [comment]
Hi @guzy0324,

as I said, you can do so inside the module computing the scale, i.e. the `param_map`.

Best regards,
Vincent
